src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -19415,6 +19415,176 @@ public void ExplicitCast_StaticType()
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (8,13): warning CS8602: Possible dereference of a null reference.
+                //             x.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(8, 13),
+                // (10,13): warning CS8602: Possible dereference of a null reference.
+                //             y.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(10, 13));
+        }
+
+        [Fact]
+        public void ForEach_03()
+        {
+            var source =
+@"class C
+{
+    static void F1(dynamic c)
+    {
+        foreach (var x in c)
+            x.ToString();
+        foreach (object? y in c)
+            y.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ForEach_04()
+        {
+            var source =
+@"using System.Collections;
+using System.Collections.Generic;
+class C<T> : IEnumerable<T> where T : class
+{
+    IEnumerator<T> IEnumerable<T>.GetEnumerator() => throw null;
+    IEnumerator IEnumerable.GetEnumerator() => throw null;
+}
+class P
+{
+    static void F<T>(C<T?> c) where T : class
+    {
+        foreach (var x in c)
+            x.ToString();
+        foreach (T? y in c)
+            y.ToString();
+        foreach (T z in c)
+            z.ToString();
+        foreach (object w in c)
+            w.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (13,13): warning CS8602: Possible dereference of a null reference.
+                //             x.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(13, 13),
+                // (15,13): warning CS8602: Possible dereference of a null reference.
+                //             y.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "y").WithLocation(15, 13),
+                // (17,13): warning CS8602: Possible dereference of a null reference.
+                //             z.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "z").WithLocation(17, 13),
+                // (19,13): warning CS8602: Possible dereference of a null reference.
+                //             w.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "w").WithLocation(19, 13));
+        }
+
+        [Fact]
+        public void ForEach_05()
+        {
+            var source =
+@"struct S<T> where T : class
+{
+    public E<T> GetEnumerator() => new E<T>();
+}
+struct E<T> where T : class
+{
+    public T Current => throw null;
+    public bool MoveNext() => false;
+}
+class P
+{
+    static void F1<T>() where T : class
```

Comment:
>T : class [](start = 30, length = 9)

Do we assume T is not nullable in this case? Could you remind why?

Reply:
The `class?` constraint is intended to indicate `T` is nullable.

---
In reply to: [186532458](https://github.com/dotnet/roslyn/pull/26660#discussion_r186532458) [](ancestors = 186532458)